internal/db/domain.go

```diff
@@ -0,0 +1,24 @@
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package db
+
+import "slices"
+
+// This file contains domain-level methods for db structs
+
+// CanImplement returns true if the provider implements the given type.
+func (p *Provider) CanImplement(impl ProviderType) bool {
+	return slices.Contains(p.Implements, impl)
+}
```

Comment:
I was just thinking about this last night.  I think this is a good first start, but I'm wondering whether it would make sense to rename `db.Provider` to `db.ProviderData` to emphasize that it's the data-at-rest representation rather than the active object representing a provider.  Either that, or we just accept that a `db.Provider` and a `provider.Provider` are unrelated to each other, and maybe use the same name more places.

sqlc renaming feature: https://docs.sqlc.dev/en/stable/howto/rename.html#tables

Reply:
> Agreed; I'd hold off on the rename until after #2841, #2842, and the follow-on that mostly hides `db.Provider` from controlplane before doing the rename.